src/communication/websocket/session.cpp

```diff
@@ -86,14 +106,68 @@ void Session::DoRead() {
       }));
 }
 
-void Session::OnRead(boost::beast::error_code ec, size_t /*bytes_transferred*/) {
+void Session::DoClose() {
+  ws_.async_close(boost::beast::websocket::close_code::normal,
+                  boost::asio::bind_executor(strand_, [shared_this = shared_from_this()](boost::beast::error_code ec) {
+                    shared_this->OnClose(ec);
+                  }));
+}
+
+void Session::OnClose(boost::beast::error_code ec) {
+  if (ec) {
+    return LogError(ec, "close");
+  }
+  connected_.store(false, std::memory_order_relaxed);
+}
+
+bool Session::Authenticate(const nlohmann::json &creds) {
+  return auth_.Authenticate(creds.at("username").get<std::string>(), creds.at("password").get<std::string>());
+}
+
+bool Session::Authorize(const nlohmann::json &creds) {
+  return auth_.DoesUserHasPermission(creds.at("username").get<std::string>(), auth::Permission::WEBSOCKET);
+}
+
+void Session::OnRead(const boost::beast::error_code ec, const size_t /*bytes_transferred*/) {
   if (ec == boost::beast::websocket::error::closed) {
     messages_.clear();
     connected_.store(false, std::memory_order_relaxed);
     return;
   }
 
-  buffer_.consume(buffer_.size());
+  if (!authenticated_) {
+    auto response = nlohmann::json();
+    try {
+      const auto creds = nlohmann::json::parse(boost::beast::buffers_to_string(buffer_.data()));
+      buffer_.consume(buffer_.size());
+
+      if (!Authenticate(creds)) {
+        response["success"] = false;
+        response["message"] = "Authentication failed!";
+        close_ = true;
+        QuickWrite(response.dump());
```

Comment:
why don't you just do 
```
MG_ASSERT(messages_.empty());
messages_.push_back(response.dump()); 
DoWrite()
```
you can even extract it into a lambda.
And the call explicitly DoClose where it needs to be called.

Reply:
Then don't forget to close the socket in that case.